core/client/src/main/java/alluxio/client/block/LocalBlockInStream.java

```diff
@@ -52,25 +53,63 @@
    * @throws IOException if I/O error occurs
    */
-  public LocalBlockInStream(long blockId, long blockSize, WorkerNetAddress workerNetAddress,
-      FileSystemContext context, InStreamOptions options) throws IOException {
+  public static LocalBlockInStream create(long blockId, long blockSize,
+      WorkerNetAddress workerNetAddress, FileSystemContext context, InStreamOptions options)
+      throws IOException {
+    Closer closer = Closer.create();
+    BlockWorkerClient client;
+    LockBlockResult result;
+    LocalFileBlockReader reader;
+    try {
+      client = closer.register(context.createBlockWorkerClient(workerNetAddress));
+      result = client.lockBlock(blockId, LockBlockOptions.defaults());
+      closer.register(new LockBlockResource(client, blockId));
+      reader = new LocalFileBlockReader(result.getBlockPath());
+    } catch (AlluxioException | IOException e) {
+      CommonUtils.closeCloserIgnoreException(closer);
+      throw CommonUtils.castToIOException(e);
+    }
+    return new LocalBlockInStream(client, blockId, blockSize, reader, options);
+  }
+
+  /**
+   * Creates a local block input stream. It requires the block to be locked before calling this.
+   *
+   * @param client the block worker client
+   * @param blockId the block id
+   * @param blockSize the size of the block
+   * @param reader the local block file reader, the {@link LocalBlockInStream} created should
+   *        close this reader
+   * @param options the instream options
+   * @return the {@link LocalBlockInStream} instance
+   */
+  public static LocalBlockInStream createWithLockedBlock(BlockWorkerClient client, long blockId,
+      long blockSize, LocalFileBlockReader reader, InStreamOptions options) {
+    return new LocalBlockInStream(client, blockId, blockSize, reader, options);
+  }
+
+  /**
+   * Creates a local block input stream. It requires the block to be locked before calling this.
+   *
+   * @param client the block worker client
+   * @param blockId the block id
+   * @param blockSize the size of the block
+   * @param reader the local block file reader which should be closed by this class
+   * @param options the instream options
+   */
+  private LocalBlockInStream(BlockWorkerClient client, long blockId,
+      long blockSize, LocalFileBlockReader reader, InStreamOptions options) {
     super(blockId, blockSize);
-    mContext = context;
+    mBlockWorkerClient = client;
+    mReader = reader;
 
     mCloser = Closer.create();
-    try {
-      mBlockWorkerClient = mCloser.register(mContext.createBlockWorkerClient(workerNetAddress));
-      LockBlockResult result = mBlockWorkerClient.lockBlock(blockId);
-      mReader = mCloser.register(new LocalFileBlockReader(result.getBlockPath()));
-    } catch (AlluxioException e) {
-      mCloser.close();
-      throw new IOException(e);
-    } catch (IOException e) {
-      mCloser.close();
-      throw e;
-    }
+    mCloser.register(client);
```

Comment:
I think the code is much more cleaner now.

here is a suggestion, which I am not strong on it, leave this up to you:  can we take `closer` as an argument from factory method passed here and assign `closer` to `mCloser`? As I see we are doing redundant work here dealing with two `closers` one in factory method and one in constr, while they are essentially the same